### PR TITLE
ZJIT: Specialize `String#byteslice` with two args

### PR DESCRIPTION
**Pull request by dak2**
## What
When the receiver is a `String` and both arguments are likely `Fixnum`, the annotation now emits a new `StringByteslice` HIR instruction that calls `rb_str_byte_substr` directly, instead of going through `CCallVariadic`.

Any other shape, one argument, a Range, or non-Fixnum arguments, keeps using CCallVariadic.

https://github.com/Shopify/ruby/issues/1036

## Bench
Measured with ruby-bench in a Docker container (aarch64-linux, --zjit), comparing master vs. this branch.
The median of 10 runs shows a consistent ~2% speedup on protoboeuf and no change on protoboeuf-encode.

```
$ ruby run_benchmarks.rb protoboeuf protoboeuf-encode \
    --chruby 'before --zjit;after --zjit'

before: ruby 4.1.0dev (2026-08-30T06:52:45Z master a889da909c) +ZJIT +PRISM [aarch64-linux]
after: ruby 4.1.0dev (2026-08-30T15:09:42Z zjit-specialize-string-byteslice-two-args 06a0bcd6fd) +ZJIT +PRISM [aarch64-linux]

-----------------  -----------  -----------  -------------  ------------
bench              before (ms)   after (ms)  after 1st itr  before/after
protoboeuf         11.5 ± 2.2%  11.1 ± 2.0%          1.022         1.035
protoboeuf-encode  13.2 ± 9.4%  13.3 ± 9.3%          0.993         0.986
-----------------  -----------  -----------  -------------  ------------
```

### CODE REVIEW

**Comment by tekknolagi** [comment]
One thing to triple check: one of the C functions, maybe this one, takes in C integers instead of VALUE. You may need to emit UnboxFixnum and pass them in that way (and fix the type checks)

**Comment by tekknolagi** [comment]
Oh it's the private function this one calls. Ok, that's fine. We will leave it as work for later to do the unboxed thing

**Comment by dak2** [comment]
@tekknolagi 
Sounds good, thanks for confirming! 
I'll keep this PR as-is and tackle unboxing beg/len in a separate follow-up PR.

File: zjit/src/hir.rs [hunk 1]
```diff
@@ -3549,6 +3561,7 @@ impl Function {
             Insn::StringIntern { .. } => types::Symbol,
             Insn::StringConcat { .. } => types::StringExact,
             Insn::StringGetbyte { .. } => types::Fixnum,
+            Insn::StringByteslice { .. } => types::StringExact.union(types::NilClass),
```
Comment (nirvdrum) [inline zjit/src/hir.rs:3564]: Can you please add a test that exercises the `nil` return value?

**Review by nirvdrum — CHANGES_REQUESTED** [review]

**Review by nirvdrum — CHANGES_REQUESTED** [review]
Thanks for the PR.  This overall looks really good. I would just like to see tests cover cases where `String#byteslice` would return `nil`.

File: zjit/src/hir.rs (hunk 1 above)
Comment (dak2) [reply zjit/src/hir.rs:3564]: @nirvdrum 
Oh, I missed it. Addressed. Thanks for the review! 
https://github.com/ruby/ruby/pull/18568/commits/25e329a1e99149ecbc0fdf82ba8a8e63befd527a

**Review by k0kubun — APPROVED** [review]
I was wondering if `rb_str_byte_substr` is indeed leaf, but it does seem leaf when two Fixnums are given. So this seems safe.

**Review by nirvdrum — APPROVED** [review]
Thanks for adding the `nil` test. This looks good to go.

File: zjit/src/hir/opt_tests.rs [hunk 2]
```diff
@@ -13015,6 +13015,113 @@ mod hir_opt_tests {
...
+        fn test@<compiled>:2:
+        bb1():
+          EntryPoint interpreter
+          v1:BasicObject = LoadSelf
+          v2:CPtr = LoadSP
+          v3:BasicObject = LoadField v2, :s@0x1000
+          v4:BasicObject = LoadField v2, :beg@0x1001
+          v5:BasicObject = LoadField v2, :len@0x1002
+          Jump bb3(v1, v3, v4, v5)
+        bb2():
```
Comment (tekknolagi) [inline zjit/src/hir/opt_tests.rs:13172]: can you add tests that this is only getting the 2 argument case? there are other cases - it's variadic, isn't it?

Comment (dak2) [reply zjit/src/hir/opt_tests.rs:13172]: @tekknolagi 

Thanks! You're right, it's variadic. 
Added tests for the one-argument and three-argument forms, which both fall back to CCallVariadic.
https://github.com/ruby/ruby/pull/18568/commits/68d204ebc1c9fd8699ea903eeb34118c7a911665

A Range argument is just the one-argument case, and since we dispatch on arity rather than the argument type, I believe the one-argument test covers it.

**Comment by dak2** [comment]
@XrXr 

Thanks for trying to merge it. The CI probably failed because of a flaky test, so I rebased it and reran it. 
Could you take another look?

**Review by rafaelfranca — APPROVED** [review]
